core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -2487,6 +2501,21 @@ private[spark] class DAGScheduler(
       executorFailureEpoch -= execId
     }
     shuffleFileLostEpoch -= execId
+
+    if (pushBasedShuffleEnabled) {
```

Comment:
Currently this is done as part of `executorAdded` event, do you think we should have a separate event which gets triggered only when the new executor added host is not already part of the existing `mergerLocs`?

Reply:
Ideally, we want to go into this `if` block only if a new host is added (since if executor is on existing/known hosts, number of mergers is not increasing) - more specifically, a new block manager in a new host is added (in `BlockManagerMasterEndpoint`).

That is why I was thinking about doing it there ...

Reply:
Agree. Let me see how we can send an event when a new blockManager's host which is not already part of existing shuffleMergers and correspondingly fire an event to DAGScheduler

Reply:
Briefly looked at this one, currently there isn't a way to communicate from `BlockManagerMasterEndpoint` to `DAGScheduler`. Also the state info is in 3 different places, `DAGScheduler` querying merger locations, `YarnSchedulerBackend` having the logic of desired num mergers and passing the `excludeNodes` info to `BlockManagerMasterEndpoint`. `BlockManagerMasterEndpoint` itself preferring active executors hosts first and the remaining merger locations from the dead executors host.

Apart from this the number of `ExecutorAdded` events should be much smaller than the `TaskStart and TaskEnd` events, so this shouldn't be any worse than what we have already. Let me know your thoughts.